flareon/src/forms.rs

```diff
@@ -0,0 +1,342 @@
+/// through the [`Form`] trait and field types that implement [`AsFormField`].
+pub trait FormField: Sized {
+    /// Custom options for the form field, unique for each field type.
+    type CustomOptions: Default;
+
+    /// Creates a new form field with the given options.
+    fn with_options(options: FormFieldOptions, custom_options: Self::CustomOptions) -> Self;
+
+    /// Returns the generic options for the form field.
+    fn options(&self) -> &FormFieldOptions;
+
+    /// Returns the ID of the form field.
+    fn id(&self) -> &str {
+        &self.options().id
+    }
+
+    /// Sets the string value of the form field.
+    ///
+    /// This method should convert the value to the appropriate type for the
+    /// field, such as a number for a number field.
+    fn set_value(&mut self, value: Cow<str>);
+
+    /// Renders the form field as an HTML string.
+    fn render(&self) -> String;
+}
+
+/// A version of [`FormField`] that can be used in a dynamic context.
+///
+/// This trait is used to allow a form field to be used in a dynamic context,
+/// such as when using Form field iterator. It provides access to the field's
+/// options, value, and rendering, among others.
+///
+/// This trait is implemented for all types that implement [`FormField`].
+pub trait DynFormField {
+    fn dyn_options(&self) -> &FormFieldOptions;
+
+    fn dyn_id(&self) -> &str;
+
+    fn dyn_set_value(&mut self, value: Cow<str>);
+
+    fn dyn_render(&self) -> String;
+}
+
+impl<T: FormField> DynFormField for T {
+    fn dyn_options(&self) -> &FormFieldOptions {
+        FormField::options(self)
+    }
+
+    fn dyn_id(&self) -> &str {
+        FormField::id(self)
+    }
+
+    fn dyn_set_value(&mut self, value: Cow<str>) {
+        FormField::set_value(self, value)
+    }
+
+    fn dyn_render(&self) -> String {
+        FormField::render(self)
+    }
+}
+
+/// A trait for types that can be used as form fields.
+///
+/// This trait uses [`FormField`] to define a type that can be used as a form
+/// field. It provides a way to clean the value of the field, which is used to
+/// validate the field's value before converting to the final type.
+pub trait AsFormField {
+    type Type: FormField;
+
+    fn clean_value(field: &Self::Type) -> Result<Self, FormFieldValidationError>
+    where
+        Self: Sized;
+}
+
+/// A form field for a string.
+#[derive(Debug)]
+pub struct CharField {
+    options: FormFieldOptions,
+    custom_options: CharFieldOptions,
+    value: Option<String>,
+}
+
+/// Custom options for a `CharField`.
+#[derive(Debug, Default)]
+pub struct CharFieldOptions {
+    /// The maximum length of the field. Used to set the `maxlength` attribute
+    /// in the HTML input element.
+    pub max_length: Option<u32>,
+}
+
+impl CharFieldOptions {
+    /// Sets the maximum length for the `CharField`.
+    pub fn set_max_length(&mut self, max_length: u32) {
+        self.max_length = Some(max_length);
+    }
+}
+
+impl FormField for CharField {
```

Comment:
This implementation will probably be moved into some macro in the next PRs as it's pretty generic, but an in-place implementation is fine as long as we only have one field type (which is the case for now).